[FG:InPlacePodVerticalScaling] kubelet: Propagate error in doPodResizeAction() to the caller

hshiina opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->
/kind bug

#### What this PR does / why we need it:
This fix makes `doPodResizeAction()` return the result instead of setting an error to the `result` argument, which should have been passed as a pointer, so that the error is propagated to the caller. Even with this fix, errors from `doPodResizeAction()` still do not surface to users (xref: #123441). So, no description will be added to the release notes.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

hshiina commented on pkg/kubelet/cm/helpers_unsupported.go:
```diff
@@ -41,7 +41,7 @@ func MilliCPUToQuota(milliCPU, period int64) int64 {
 
 // MilliCPUToShares converts the milliCPU to CFS shares.
-func MilliCPUToShares(milliCPU int64) int64 {
+func MilliCPUToShares(milliCPU int64) uint64 {
```
This is being fixed separately in #127293. After it is merged, I will rebase this PR.

ffromani commented:
/cc @esotsal 

/ok-to-test

ffromani commented:
/cc @vinaykul

haircommander commented:
/triage accepted
/priority important-longterm

ffromani reviewed (commented):
I'm not sure about mockery. The use of mocks is sparse and I didn't find precise guidelines (but I haven't ever looked too hard either). That's pretty minor though.

esotsal commented:
I agree with @ffromani comment, @hshiina can the mockery related comment be excluded from this PR and keep only refactoring for this moment? We could create a separate PR for introducing mockery in InPlacePodVerticalScaling testing in the future perhaps.

esotsal commented:
/assign

hshiina commented:
/test pull-kubernetes-node-kubelet-serial-podresize

hshiina commented:
/retest

esotsal commented:
Thanks @hshiina ! /lgtm

esotsal commented:
/cc @tallclair

googs1025 commented:
@hshiina 
So do I need to move the changes from this PR to this PR https://github.com/kubernetes/kubernetes/pull/127184?

hshiina commented:
> @hshiina So do I need to move the changes from this PR to this PR #127184?

If you import #125654 into your PR (#127184) in order to emit events for failures in `doPodResizeAction()`, please do so. #125654 depended on this PR (#127300) to get errors from `doPodResizeAction()`.

esotsal commented:
/assign @tallclair

esotsal commented:
@hshiina should we propagate also error here? 

https://github.com/kubernetes/kubernetes/blob/91cab7cea0afcc607214a4e24752a662a951c296/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L934C1-L940C3

vinaykul commented:
Thanks for the PR. Addition of the missed unit test is certainly helpful. To clarify, the PR in its current does not fix https://github.com/kubernetes/kubernetes/issues/123441, does it? (which I guess is due to [minikube cri-dockerd not reporting ContainerResources in ContainerStatus](https://github.com/kubernetes/kubernetes/pull/128128#issuecomment-2417826822))

vinaykul commented:
/test pull-kubernetes-e2e-inplace-pod-resize-containerd-main-v2

hshiina commented:
> Thanks for the PR. Addition of the missed unit test is certainly helpful. To clarify, the PR in its current does not fix #123441, does it? (which I guess is due to [minikube cri-dockerd not reporting ContainerResources in ContainerStatus](https://github.com/kubernetes/kubernetes/pull/128128#issuecomment-2417826822))

No, this PR does not fix #123441. I mentioned this issue because I guessed it was caused by an error in `podResizeAction()`. Even if it's true, another work like #125654 and #127184 to emit an event based on an error propagated by this fix is necessary.

vinaykul commented:
> > Thanks for the PR. Addition of the missed unit test is certainly helpful. To clarify, the PR in its current does not fix #123441, does it? (which I guess is due to [minikube cri-dockerd not reporting ContainerResources in ContainerStatus](https://github.com/kubernetes/kubernetes/pull/128128#issuecomment-2417826822))
> 
> No, this PR does not fix #123441. I mentioned this issue because I guessed it was caused by an error in `podResizeAction()`. Even if it's true, another work like #125654 and #127184 to emit an event based on an error propagated by this fix is necessary.

Ack. Let's use this opportunity to make the error propagation more precise and consistent with Create|Kill-container/Create|Kill-podsandbox. Please add [two SyncActions](https://github.com/kubernetes/kubernetes/blob/9d967ff97332a024b8ae5ba89c83c239474f42fd/pkg/kubelet/container/sync_result.go#L54-L71) for UpdateContainer & UpdatePodSandbox, define [two new errors](https://github.com/kubernetes/kubernetes/blob/9d967ff97332a024b8ae5ba89c83c239474f42fd/pkg/kubelet/container/sync_result.go#L37-L48) associated with them, and follow the [killPodwithSyncResult](https://github.com/kubernetes/kubernetes/blob/9d967ff97332a024b8ae5ba89c83c239474f42fd/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L1467-L1485) pattern for doPodResizeAction. Thanks.

vinaykul commented:
It also looks like pull-kubernetes-e2e-inplace-pod-resize-containerd-main-v2 job is failing a lot: https://prow.k8s.io/?job=pull-kubernetes-e2e-inplace-pod-resize-containerd-main-v2 .. the scheduler-focussed test mostly but there is some other flakiness.  Not related to this change though, but it needs to be fixed.

dshebib commented:
@hshiina It seems like this may be a more complicated change. Would you prefer to work on this and have the event emission as a follow-up, or should they be in the same PR?

hshiina commented:
> Ack. Let's use this opportunity to make the error propagation more precise and consistent with Create|Kill-container/Create|Kill-podsandbox. Please add [two SyncActions](https://github.com/kubernetes/kubernetes/blob/9d967ff97332a024b8ae5ba89c83c239474f42fd/pkg/kubelet/container/sync_result.go#L54-L71) for UpdateContainer & UpdatePodSandbox, define [two new errors](https://github.com/kubernetes/kubernetes/blob/9d967ff97332a024b8ae5ba89c83c239474f42fd/pkg/kubelet/container/sync_result.go#L37-L48) associated with them, and follow the [killPodwithSyncResult](https://github.com/kubernetes/kubernetes/blob/9d967ff97332a024b8ae5ba89c83c239474f42fd/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L1467-L1485) pattern for doPodResizeAction. Thanks.

#127479 is adopting this approach. It might be better to continue this work on #127479.
/hold

tallclair commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -725,49 +724,42 @@ func (m *kubeGenericRuntimeManager) doPodResizeAction(pod *v1.Pod, podStatus *ku
 	if len(podContainerChanges.ContainersToUpdate[v1.ResourceMemory]) > 0 || podContainerChanges.UpdatePodResources {
 		if podResources.Memory == nil {
 			klog.ErrorS(nil, "podResources.Memory is nil", "pod", pod.Name)
```
delete this

tallclair commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -662,14 +662,13 @@ func (m *kubeGenericRuntimeManager) computePodResizeAction(pod *v1.Pod, containe
 	//TODO(vinaykul,InPlacePodVerticalScaling): Figure out best way to get enforceMemoryQoS value (parameter #4 below) in platform-agnostic way
 	podResources := cm.ResourceConfigForPod(pod, m.cpuCFSQuota, uint64((m.cpuCFSQuotaPeriod.Duration)/time.Microsecond), false)
 	if podResources == nil {
 		klog.ErrorS(nil, "Unable to get resource configuration", "pod", pod.Name)
```
delete this

tallclair commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -662,14 +662,13 @@ func (m *kubeGenericRuntimeManager) computePodResizeAction(pod *v1.Pod, containe
 		klog.ErrorS(nil, "Unable to get resource configuration", "pod", pod.Name)
-		result.Fail(fmt.Errorf("Unable to get resource configuration processing resize for pod %s", pod.Name))
-		return
+		return fmt.Errorf("unable to get resource configuration processing resize for pod %s", pod.Name)
```
```suggestion
		return fmt.Errorf("unable to get resource configuration processing resize for pod %s", format.Pod(pod))
```

tallclair commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -725,49 +724,42 @@ func (m *kubeGenericRuntimeManager) doPodResizeAction(pod *v1.Pod, podStatus *ku
 			klog.ErrorS(nil, "podResources.Memory is nil", "pod", pod.Name)
-			result.Fail(fmt.Errorf("podResources.Memory is nil for pod %s", pod.Name))
-			return
+			return fmt.Errorf("podResources.Memory is nil for pod %s", pod.Name)
```
```suggestion
			return fmt.Errorf("podResources.Memory is nil for pod %s", format.Pod(pod))
```

tallclair commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -725,49 +724,42 @@ func (m *kubeGenericRuntimeManager) doPodResizeAction(pod *v1.Pod, podStatus *ku
 		}
 		currentPodMemoryConfig, err := pcm.GetPodCgroupConfig(pod, v1.ResourceMemory)
 		if err != nil {
 			klog.ErrorS(err, "GetPodCgroupConfig for memory failed", "pod", pod.Name)
```
Add this context to the returned error instead

tallclair commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -725,49 +724,42 @@ func (m *kubeGenericRuntimeManager) doPodResizeAction(pod *v1.Pod, podStatus *ku
 		}
 		currentPodMemoryUsage, err := pcm.GetPodCgroupMemoryUsage(pod)
 		if err != nil {
 			klog.ErrorS(err, "GetPodCgroupMemoryUsage failed", "pod", pod.Name)
```
here too, add context to return, and delete the log statement.

tallclair commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -725,49 +724,42 @@ func (m *kubeGenericRuntimeManager) doPodResizeAction(pod *v1.Pod, podStatus *ku
+			return err
 		}
 		if currentPodMemoryUsage >= uint64(*podResources.Memory) {
 			klog.ErrorS(nil, "Aborting attempt to set pod memory limit less than current memory usage", "pod", pod.Name)
```
delete the log

tallclair commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -725,49 +724,42 @@ func (m *kubeGenericRuntimeManager) doPodResizeAction(pod *v1.Pod, podStatus *ku
 			klog.ErrorS(nil, "Aborting attempt to set pod memory limit less than current memory usage", "pod", pod.Name)
-			result.Fail(fmt.Errorf("Aborting attempt to set pod memory limit less than current memory usage for pod %s", pod.Name))
-			return
+			return fmt.Errorf("aborting attempt to set pod memory limit less than current memory usage for pod %s", pod.Name)
```
```suggestion
			return fmt.Errorf("aborting attempt to set pod memory limit less than current memory usage for pod %s", format.Pod(pod))
```

tallclair commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -725,49 +724,42 @@ func (m *kubeGenericRuntimeManager) doPodResizeAction(pod *v1.Pod, podStatus *ku
 	}
 	if len(podContainerChanges.ContainersToUpdate[v1.ResourceCPU]) > 0 || podContainerChanges.UpdatePodResources {
 		if podResources.CPUQuota == nil || podResources.CPUShares == nil {
 			klog.ErrorS(nil, "podResources.CPUQuota or podResources.CPUShares is nil", "pod", pod.Name)
```
delete

tallclair commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -725,49 +724,42 @@ func (m *kubeGenericRuntimeManager) doPodResizeAction(pod *v1.Pod, podStatus *ku
 		}
 		currentPodCpuConfig, err := pcm.GetPodCgroupConfig(pod, v1.ResourceCPU)
 		if err != nil {
 			klog.ErrorS(err, "GetPodCgroupConfig for CPU failed", "pod", pod.Name)
```
delete log & add context to return

tallclair commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -725,49 +724,42 @@ func (m *kubeGenericRuntimeManager) doPodResizeAction(pod *v1.Pod, podStatus *ku
 			int64(*currentPodCpuConfig.CPUShares), int64(*podResources.CPUShares)); errResize != nil {
-			result.Fail(errResize)
-			return
+			return errResize
```
Add context to this error.

tallclair commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -1320,7 +1312,9 @@ func (m *kubeGenericRuntimeManager) SyncPod(ctx context.Context, pod *v1.Pod, po
 		if len(podContainerChanges.ContainersToUpdate) > 0 || podContainerChanges.UpdatePodResources {
-			m.doPodResizeAction(pod, podStatus, podContainerChanges, result)
+			if err := m.doPodResizeAction(pod, podStatus, podContainerChanges); err != nil {
+				result.Fail(err)
```
Log the error here too

tallclair commented on pkg/kubelet/kuberuntime/kuberuntime_manager_test.go:
```diff
@@ -2745,3 +2747,115 @@ func TestGetImageVolumes(t *testing.T) {
 	}
 }
+
+// This test focuses on verifying `doPodResizeAction()` propagates an error correctly with a few error cases for PR #127300.
```
```suggestion
// This test focuses on verifying `doPodResizeAction()` propagates an error correctly with a few error cases.
```

No need to add the PR context, that's what git blame is for

tallclair commented on pkg/kubelet/kuberuntime/kuberuntime_manager_test.go:
```diff
@@ -2745,3 +2747,115 @@ func TestGetImageVolumes(t *testing.T) {
+				Limits:   v1.ResourceList{v1.ResourceCPU: cpu200m, v1.ResourceMemory: mem200M},
+			},
+			qosClass:      v1.PodQOSGuaranteed,
+			expectedError: "not implemented",
```
Where is this error coming from? The fake runtime? Add a comment to explain.

vinaykul commented:
> > Ack. Let's use this opportunity to make the error propagation more precise and consistent with Create|Kill-container/Create|Kill-podsandbox. Please add [two SyncActions](https://github.com/kubernetes/kubernetes/blob/9d967ff97332a024b8ae5ba89c83c239474f42fd/pkg/kubelet/container/sync_result.go#L54-L71) for UpdateContainer & UpdatePodSandbox, define [two new errors](https://github.com/kubernetes/kubernetes/blob/9d967ff97332a024b8ae5ba89c83c239474f42fd/pkg/kubelet/container/sync_result.go#L37-L48) associated with them, and follow the [killPodwithSyncResult](https://github.com/kubernetes/kubernetes/blob/9d967ff97332a024b8ae5ba89c83c239474f42fd/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L1467-L1485) pattern for doPodResizeAction. Thanks.
> 
> #127479 is adopting this approach. It might be better to continue this work on #127479. /hold

@hshiina Hope you don't mind. But as I started thinking holistically about this in context of PRs https://github.com/kubernetes/kubernetes/pull/127479 and https://github.com/kubernetes/kubernetes/pull/127184 , I wrote some code to flesh out what I have in mind to model it consistent with killPodWithSyncResult . Please feel free to pick up the below patch and run with it. The change includes some of the changes @tallclair requested, and seems to work, but needs more testing and unit tests.

```go
diff --git a/pkg/kubelet/container/sync_result.go b/pkg/kubelet/container/sync_result.go
index 39ab04a3cac..f243f389c7d 100644
--- a/pkg/kubelet/container/sync_result.go
+++ b/pkg/kubelet/container/sync_result.go
@@ -45,6 +45,12 @@ var (
        ErrConfigPodSandbox = errors.New("ConfigPodSandboxError")
        // ErrKillPodSandbox returned when runtime failed to stop pod's sandbox.
        ErrKillPodSandbox = errors.New("KillPodSandboxError")
+       // ErrUpdatePodSandbox returned when runtime failed to update the pod's sandbox config.
+       ErrUpdatePodSandbox = errors.New("UpdatePodSandboxError")
+       // ErrUpdateContainerMemory returned when runtime failed to update the pod's container config.
+       ErrUpdateContainerMemory = errors.New("UpdateContainerMemoryError")
+       // ErrUpdateContainerCPU returned when runtime failed to update the pod's container config.
+       ErrUpdateContainerCPU = errors.New("UpdateContainerCPUError")
 )
 
 // SyncAction indicates different kind of actions in SyncPod() and KillPod(). Now there are only actions
@@ -68,6 +74,12 @@ const (
        ConfigPodSandbox SyncAction = "ConfigPodSandbox"
        // KillPodSandbox action
        KillPodSandbox SyncAction = "KillPodSandbox"
+       // UpdatePodSandbox action
+       UpdatePodSandbox SyncAction = "UpdatePodSandbox"
+       // UpdateContainerMemory action
+       UpdateContainerMemory SyncAction = "UpdateContainerMemory"
+       // UpdateContainerCPU action
+       UpdateContainerCPU SyncAction = "UpdateContainerCPU"
 )
 
 // SyncResult is the result of sync action.
diff --git a/pkg/kubelet/kuberuntime/kuberuntime_container.go b/pkg/kubelet/kuberuntime/kuberuntime_container.go
index 2a1ef3d15cd..80d12aefb4c 100644
--- a/pkg/kubelet/kuberuntime/kuberuntime_container.go
+++ b/pkg/kubelet/kuberuntime/kuberuntime_container.go
@@ -391,12 +391,11 @@ func (m *kubeGenericRuntimeManager) generateContainerConfig(ctx context.Context,
        return config, cleanupAction, nil
 }
 
-func (m *kubeGenericRuntimeManager) updateContainerResources(pod *v1.Pod, container *v1.Container, containerID kubecontainer.ContainerID) error {
+func (m *kubeGenericRuntimeManager) updateContainerResources(ctx context.Context, pod *v1.Pod, container *v1.Container, containerID kubecontainer.ContainerID) error {
        containerResources := m.generateContainerResources(pod, container)
        if containerResources == nil {
                return fmt.Errorf("container %q updateContainerResources failed: cannot generate resources config", containerID.String())
        }
-       ctx := context.Background()
        err := m.runtimeService.UpdateContainerResources(ctx, containerID.ID, containerResources)
        if err != nil {
                klog.ErrorS(err, "UpdateContainerResources failed", "container", containerID.String())
diff --git a/pkg/kubelet/kuberuntime/kuberuntime_manager.go b/pkg/kubelet/kuberuntime/kuberuntime_manager.go
index 0c7a9b24c63..3936357a489 100644
--- a/pkg/kubelet/kuberuntime/kuberuntime_manager.go
+++ b/pkg/kubelet/kuberuntime/kuberuntime_manager.go
@@ -650,7 +650,9 @@ func (m *kubeGenericRuntimeManager) computePodResizeAction(pod *v1.Pod, containe
        return true
 }
 
-func (m *kubeGenericRuntimeManager) doPodResizeAction(pod *v1.Pod, podStatus *kubecontainer.PodStatus, podContainerChanges podActions, result kubecontainer.PodSyncResult) {
+func (m *kubeGenericRuntimeManager) doPodResizeAction(ctx context.Context, pod *v1.Pod, podStatus *kubecontainer.PodStatus, podContainerChanges podActions) (result kubecontainer.PodSyncResult) {
+       updatePodResult := kubecontainer.NewSyncResult(kubecontainer.UpdatePodSandbox, pod.UID)
+       result.AddSyncResult(updatePodResult)
        pcm := m.containerManager.NewPodContainerManager()
        //TODO(vinaykul,InPlacePodVerticalScaling): Figure out best way to get enforceMemoryQoS value (parameter #4 below) in platform-agnostic way
        podResources := cm.ResourceConfigForPod(pod, m.cpuCFSQuota, uint64((m.cpuCFSQuotaPeriod.Duration)/time.Microsecond), false)
@@ -696,9 +698,13 @@ func (m *kubeGenericRuntimeManager) doPodResizeAction(pod *v1.Pod, podStatus *ku
                        }
                }
                if len(podContainerChanges.ContainersToUpdate[rName]) > 0 {
-                       if err = m.updatePodContainerResources(pod, rName, podContainerChanges.ContainersToUpdate[rName]); err != nil {
-                               klog.ErrorS(err, "updatePodContainerResources failed", "pod", format.Pod(pod), "resource", rName)
-                               return err
+                       updateContainerResults, errUpdate := m.updatePodContainerResources(ctx, pod, rName, podContainerChanges.ContainersToUpdate[rName])
+                       for _, containerResult := range updateContainerResults {
+                               result.AddSyncResult(containerResult)
+                       }
+                       if errUpdate != nil {
+                               klog.ErrorS(errUpdate, "updatePodContainerResources failed", "pod", format.Pod(pod), "resource", rName)
+                               return errUpdate
                        }
                }
                // At downsizing, requests should shrink prior to limits in order to keep "requests <= limits".
@@ -760,17 +766,20 @@ func (m *kubeGenericRuntimeManager) doPodResizeAction(pod *v1.Pod, podStatus *ku
                        return
                }
        }
+       return
 }
 
-func (m *kubeGenericRuntimeManager) updatePodContainerResources(pod *v1.Pod, resourceName v1.ResourceName, containersToUpdate []containerToUpdateInfo) error {
+func (m *kubeGenericRuntimeManager) updatePodContainerResources(ctx context.Context, pod *v1.Pod, resourceName v1.ResourceName, containersToUpdate []containerToUpdateInfo) (updateResults []*kubecontainer.SyncResult, err error) {
        klog.V(5).InfoS("Updating container resources", "pod", klog.KObj(pod))
 
        for _, cInfo := range containersToUpdate {
+               var updateContainerResult *kubecontainer.SyncResult
                container := pod.Spec.Containers[cInfo.apiContainerIdx].DeepCopy()
                // If updating memory limit, use most recently configured CPU request and limit values.
                // If updating CPU request and limit, use most recently configured memory request and limit values.
                switch resourceName {
                case v1.ResourceMemory:
+                       updateContainerResult = kubecontainer.NewSyncResult(kubecontainer.UpdateContainerMemory, container.Name)
                        container.Resources.Limits = v1.ResourceList{
                                v1.ResourceCPU:    *resource.NewMilliQuantity(cInfo.currentContainerResources.cpuLimit, resource.DecimalSI),
                                v1.ResourceMemory: *resource.NewQuantity(cInfo.desiredContainerResources.memoryLimit, resource.BinarySI),
@@ -780,6 +789,7 @@ func (m *kubeGenericRuntimeManager) updatePodContainerResources(pod *v1.Pod, res
                                v1.ResourceMemory: *resource.NewQuantity(cInfo.desiredContainerResources.memoryRequest, resource.BinarySI),
                        }
                case v1.ResourceCPU:
+                       updateContainerResult = kubecontainer.NewSyncResult(kubecontainer.UpdateContainerCPU, container.Name)
                        container.Resources.Limits = v1.ResourceList{
                                v1.ResourceCPU:    *resource.NewMilliQuantity(cInfo.desiredContainerResources.cpuLimit, resource.DecimalSI),
                                v1.ResourceMemory: *resource.NewQuantity(cInfo.currentContainerResources.memoryLimit, resource.BinarySI),
@@ -789,12 +799,19 @@ func (m *kubeGenericRuntimeManager) updatePodContainerResources(pod *v1.Pod, res
                                v1.ResourceMemory: *resource.NewQuantity(cInfo.currentContainerResources.memoryRequest, resource.BinarySI),
                        }
                }
-               if err := m.updateContainerResources(pod, container, cInfo.kubeContainerID); err != nil {
+               updateResults = append(updateResults, updateContainerResult)
+               if err = m.updateContainerResources(ctx, pod, container, cInfo.kubeContainerID); err != nil {
                        // Log error and abort as container updates need to succeed in the order determined by computePodResizeAction.
                        // The recovery path is for SyncPod to keep retrying at later times until it succeeds.
                        klog.ErrorS(err, "updateContainerResources failed", "container", container.Name, "cID", cInfo.kubeContainerID,
                                "pod", format.Pod(pod), "resourceName", resourceName)
-                       return err
+                       switch resourceName {
+                       case v1.ResourceMemory:
+                               updateContainerResult.Fail(kubecontainer.ErrUpdateContainerMemory, err.Error())
+                       case v1.ResourceCPU:
+                               updateContainerResult.Fail(kubecontainer.ErrUpdateContainerCPU, err.Error())
+                       }
+                       return
                }
                // If UpdateContainerResources is error-free, it means desired values for 'resourceName' was accepted by runtime.
                // So we update currentContainerResources for 'resourceName', which is our view of most recently configured resources.
@@ -808,7 +825,7 @@ func (m *kubeGenericRuntimeManager) updatePodContainerResources(pod *v1.Pod, res
                        cInfo.currentContainerResources.cpuRequest = cInfo.desiredContainerResources.cpuRequest
                }
        }
-       return nil
+       return
 }
 
 // computePodActions checks whether the pod spec has changed and returns the changes if true.
@@ -1308,7 +1325,11 @@ func (m *kubeGenericRuntimeManager) SyncPod(ctx context.Context, pod *v1.Pod, po
        // Step 7: For containers in podContainerChanges.ContainersToUpdate[CPU,Memory] list, invoke UpdateContainerResources
        if IsInPlacePodVerticalScalingAllowed(pod) {
                if len(podContainerChanges.ContainersToUpdate) > 0 || podContainerChanges.UpdatePodResources {
-                       m.doPodResizeAction(pod, podStatus, podContainerChanges, result)
+                       resizeResult := m.doPodResizeAction(ctx, pod, podStatus, podContainerChanges)
+                       result.AddPodSyncResult(resizeResult)
+                       if resizeResult.Error() != nil {
+                               klog.ErrorS(resizeResult.Error(), "doPodResizeAction failed")
+                       }
                }
        }
 
```

hshiina committed:
kubelet: Propagate error in doPodResizeAction() to the caller

This fix makes doPodResizeAction() return the result instead of setting
an error to the `result` argument, which should have been passed as a
pointer, so that the error is propagated to the caller. This fix also
makes the usage of PodSyncResult more consistent with other operations
like starting and killing a container.

hshiina commented:
Thank you for the feedback. I updated this PR.
/unhold

tallclair commented:
Can you help me understand the impact of this change? Is this mostly just a cleanup, or does this result in errors being handled differently?

vinaykul commented:
> Can you help me understand the impact of this change? Is this mostly just a cleanup, or does this result in errors being handled differently?

It adds more precise logging and any associated errors for the action in a manner similar to what is done in killPodWithSyncResult. e.g from my quick-n-dirty test kubelet log.
```bash
E1104 06:40:39.043476  589620 kubelet.go:1988] DBG: SR: &{Action:UpdatePodSandbox Target:bbdfbdad-fd0d-47ec-8666-3ce74d260256 Error:<nil> Message:}
E1104 06:40:39.043480  589620 kubelet.go:1988] DBG: SR: &{Action:UpdateContainerMemory Target:ctr Error:<nil> Message:}
E1104 06:40:39.043484  589620 kubelet.go:1988] DBG: SR: &{Action:UpdateContainerCPU Target:ctr Error:<nil> Message:}
```

vinaykul commented:
/lgtm

tallclair commented:
/approve

pacoxu commented:
/milestone v1.32
approved before code freeze

tallclair commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -673,10 +674,7 @@ func (m *kubeGenericRuntimeManager) doPodResizeAction(pod *v1.Pod, podStatus *ku
-			klog.ErrorS(err, "Failed to set cgroup config", "resource", rName, "pod", pod.Name)
-		}
-		return err
+		return fmt.Errorf("failed to set cgroup config for %s of pod %s: %w", rName, format.Pod(pod), err)
```
This is unconditionally returning an error for this function.

hshiina replied:
I'm sorry I forgot to run tests after updating.

vinaykul-ai replied:
I believe the presubmit job is run when these files are touched, but I overlooked checking it on this instance, and missed the bug in code review as well - my bad (note to self: dont review code with tired eyes)
Imho this may be a good time to make that presubmit job visible.

vinaykul replied:
I believe the e2e tests presubmit job runs when these files are touched. But I overlooked checking for the outcome on this occasion and missed the bug in the review as well, my bad.
Imho this is a good time to make pull-kubernetes-e2e-inplace-pod-resize-containerd-main-v2 visible in PRs.

vinaykul replied:
PR https://github.com/kubernetes/test-infra/pull/33781
